packages/@lwc/ssr-runtime/src/reflection.ts

```diff
@@ -34,27 +134,33 @@ export function reflectAttrToProp(
     }
 }
 
-export const descriptors = create(null);
-for (const [attrName, propName] of entries(attrsToProps)) {
-    descriptors[propName] = {
-        get(this: LightningElement): string | null {
-            return this.getAttribute(attrName);
+export const descriptors: Record<string, PropertyDescriptor> = {
+    accessKey: stringDescriptor('accesskey'),
+    dir: stringDescriptor('dir'),
+    draggable: explicitBooleanDescriptor('draggable', true),
+    hidden: booleanAttributeDescriptor('hidden'),
+    id: stringDescriptor('id'),
+    lang: stringDescriptor('lang'),
+    spellcheck: explicitBooleanDescriptor('spellcheck', false),
+    tabIndex: {
+        get(this: LightningElement): number {
+            const str = this.getAttribute('tabindex');
+            const num = Number(str);
+            return isFinite(num) ? Math.trunc(num) : -1;
         },
-        set(this: LightningElement, newValue: unknown): void {
-            const currentValue = this.getAttribute(attrName);
-            if (newValue !== currentValue) {
-                // TODO [#3284]: According to the spec, IDL nullable type values
-                // (null and undefined) should remove the attribute; however, we
-                // only do so in the case of null for historical reasons.
-                // See also https://github.com/w3c/aria/issues/1858
-                if (isNull(newValue)) {
-                    this.removeAttribute(attrName);
-                } else {
-                    this.setAttribute(attrName, toString(newValue));
-                }
+        set(this: LightningElement, newValue: number): void {
+            const currentValue = this.getAttribute('tabindex');
+            const num = Number(newValue);
+            const normalizedValue = isFinite(num) ? String(Math.trunc(num)) : '0';
```

Comment:
LWC enforces that `tabindex` in templates is only `0` or `-1`. For string attributes (e.g .`tabindex="123"`), it throws a compile time error for invalid values. For dynamic attributes (e.g. `tabindex={prop}`), it caps the maximum value at `0`. Values less than `-1` are still permitted. More relevant to this PR, the value used when setting `this.tabIndex = something` is not modified.

[playground](https://stackblitz.com/edit/salesforce-lwc-wbzinf?file=src%2Fmodules%2Fx%2Fcounter%2Fcounter.js,src%2Fmodules%2Fx%2Fcounter%2Fcounter.html&title=LWC%20playground)